### PR TITLE
Adding cursor coordinates to TextInput

### PR DESCRIPTION
**Pull request by perunt**
## Summary:

This pull request aims to address the lack of information about the cursor (caret) position in the TextInput component. Currently, only the number of symbols where the cursor is located can be obtained(start and end cursor position), but this PR introduces a new property to track the cursor position in the x-y system coordinates.
iOS - https://github.com/facebook/react-native/pull/38110
Android - https://github.com/facebook/react-native/pull/38123
## Changelog:

[GENERAL] [ADDED] - To achieve this, a new property has been added to track the cursor position, and the initializer has been updated to accept it. The conversion method has also been modified to parse this property from JSON and pass it to the updated initializer. This allows the code to handle cursor position alongside text selection information.

With this PR, the onSelectionChange method will now return the cursor position alongside the start and end positions, as follows:
```
selection: {
    start: number,
    end: number,
    cursorPosition: {
        start: {x: number, y: number},
        end: {x: number, y: number}
    }
}
```
The cursorPosition object contains the coordinates of the top left corner (start) and the bottom right corner (end) of the cursor or selected text.

### CODE REVIEW

**Review by cortinico — CHANGES_REQUESTED** [review]
Hey @perunt 
Can you please take a look at the CI signal as the Android code is not compiling

**Comment by perunt** [comment]
Now it's compiling successfully, and all tests are passing well. Please let me know if you notice any other issues.

**Comment by acoates-ms** [comment]
Can you add a RNTester page or add to an exisiting one, a usage of the new API?
I dont see the new properties added to the type information in the JS?  I'd expect to see an update to TextInput.js's Selection type, and a matching change in TextInput.d.ts.

**Comment by acoates-ms** [comment]
I suspect this also needs some changes in fabric to support the extra properties?

**Comment by NickGerleman** [comment]
Yes, I think previously the RN team had decided we will not be taking new props with only a Paper implementation, since it means these will stop working the moment we switch to Fabric as the default.

**Comment by NickGerleman** [comment]
I would also expect this series of changes to include both Flow and TypeScript types. Right now using this API would require suppressing typecheking errors.

**Comment by perunt** [comment]
Got it, let me address all this stuff

**Comment by PPatBoyd** [comment]
Would it make sense for the caret position to be a pair of points / rect?  What point does the current XY represent?

File: packages/react-native/ReactAndroid/src/main/java/com/facebook/react/views/textinput/ReactTextInputSelectionEvent.java [hunk 1]
```diff
@@ -20,17 +20,31 @@
...
   @Deprecated
-  public ReactTextInputSelectionEvent(int viewId, int selectionStart, int selectionEnd) {
-    this(ViewUtil.NO_SURFACE_ID, viewId, selectionStart, selectionEnd);
+  public ReactTextInputSelectionEvent(
+      int viewId,
+      int selectionStart,
+      int selectionEnd,
+      float cursorPositionX,
+      float cursorPositionY) {
+    this(-1, viewId, selectionStart, selectionEnd, cursorPositionX, cursorPositionY);
```
[the remark this answers is not in the document]
Comment (perunt) [reply packages/react-native/ReactAndroid/src/main/java/com/facebook/react/views/textinput/ReactTextInputSelectionEvent.java:33]: I used an old code version before `NO_SURFACE_ID` was implemented (https://github.com/facebook/react-native/commit/d2d2a8a598cb8b5e5218c3c21cef408de2142fe4)
I'll update it

**Comment by perunt** [comment]
The XY is just pointing to the top-left start of the selection right now. I get what you're saying - it might be a good move to add the end point too.

**Comment by acoates-ms** [comment]
> The XY is just pointing to the top-left start of the selection right now. I get what you're saying - it might be a good move to add the end point too.

I'm not sure what you are doing with these coords, but many of the use cases I can think of would require having both topleft and bottom right coords.  Want to show a menu at the cursor? - depending on if the menu is going to show above or below the cursor, you'd need both coords.

**Comment by perunt** [comment]
> I'm not sure what you are doing with these coords, but many of the use cases I can think of would require having both topleft and bottom right coords. Want to show a menu at the cursor? - depending on if the menu is going to show above or below the cursor, you'd need both coords.

Yes, my goal is to use this for menu positioning. I've attempted using both coordinates as you've suggested, but encountered a bit of a hiccup when it came to selection. I've been considering whether it should just pinpoint to a single location instead.
I can do the two-coordinates thing. Do you reckon I should do the same with the selection? I mean, if there were a few lines of selection, it would show the top left and bottom right corner coordinates,

**Comment by perunt** [comment]
@Szymon20000 will appreciate any thoughts on it, thanks!

**Review by NickGerleman — CHANGES_REQUESTED** [review]
It looks like this does not implement the change on iOS Fabric at least (did a quick check that `RCTTextInputComponentView` does not reuse the view edited).

This change seems fairly mechanical, but we usually ask folks to split these sorts of PRs, to enable finding the most appropriate reviewers for each part (and to make it easy to ship things as they become ready). Right now we have:
1. New event data, and new prop
2. Android, iOS, and JS layer changes
3. Paper and Fabric support (these may end up being the same on Android)

If this change gets much larger, I think we will want to split this up a bit, and might actually recommend we do that for the next iteration (e.g. separate PRs for Android and iOS).

I have not looked closely at the implementation for this yet, but the other questions I had were around the new API to allow explicitly controlling cursor position. I don't see this mentioned in your description, but I am curious how explicit control of cursor position interacts with character index based selection. Are these independent?

**Comment by perunt** [comment]
> I have not looked closely at the implementation for this yet, but the other questions I had were around the new API to allow explicitly controlling cursor position. I don't see this mentioned in your description, but I am curious how explicit control of cursor position interacts with character index based selection. Are these independent?

Thanks for the feedback! So, about that cursor position - It updates when the `onSelectionChanged` event fires and it doesn't refresh when we explicitly set the selection. 
Now, about the Fabric support, yep, I dropped the ball there. I was testing on a fork from the main and completely spaced on checking if changing the flag really turned Fabric. Good catch! I'm on it - will set up new PRs for Android and iOS Fabric support asap.

**Comment by perunt** [comment]
iOS - In this pr (https://github.com/facebook/react-native/pull/38110), I've added Fabric support and fixed the coordinates update when controlling cursor position explicitly.

Android - In this pr (https://github.com/facebook/react-native/pull/38123), It works with Fabric support and has the same updates for coordinates.

I've separated these for iOS and Android for better code review. However, I'm unsure how to handle the JavaScript portion as it's part of the example. @NickGerleman should I create separate pull requests for iOS fabric and paper?

**Comment by perunt** [comment]
@NickGerleman all necessary updates made in that PRs

> iOS - In this pr (https://github.com/facebook/react-native/pull/38110), I've added Fabric support and fixed the coordinates update when controlling cursor position explicitly.
Android - In this pr (https://github.com/facebook/react-native/pull/38123), It works with Fabric support and has the same updates for coordinates.

Current PR I'll leave for grouping this issue

**Pull request closed without merging**